crates/ruff_python_formatter/src/lib.rs

```diff
@@ -3,23 +3,85 @@ use rustpython_parser::ast::Mod;
 use rustpython_parser::lexer::lex;
 use rustpython_parser::{parse_tokens, Mode};
 
+use ruff_formatter::formatter::Formatter;
+use ruff_formatter::prelude::source_position;
 use ruff_formatter::{
-    format, FormatResult, Formatted, IndentStyle, Printed, SimpleFormatOptions, SourceCode,
+    format, write, Buffer, FormatResult, Formatted, IndentStyle, Printed, SimpleFormatOptions,
+    SourceCode,
 };
+use ruff_python_ast::node::AstNode;
 use ruff_python_ast::source_code::{CommentRanges, CommentRangesBuilder, Locator};
 
 use crate::comments::Comments;
-use crate::context::ASTFormatContext;
+use crate::context::PyFormatContext;
 use crate::module::FormatModule;
 
 pub mod cli;
 mod comments;
-pub mod context;
-mod module;
+pub(crate) mod context;
+pub(crate) mod expression;
+pub(crate) mod module;
+pub(crate) mod other;
+pub(crate) mod pattern;
 mod prelude;
+pub(crate) mod statement;
 
 include!("../../ruff_formatter/shared_traits.rs");
 
+pub(crate) type PyFormatter<'buf, 'ast> = Formatter<'buf, PyFormatContext<'ast>>;
+
+/// Rule for formatting a JavaScript [`AstNode`].
+pub(crate) trait FormatNodeRule<N>
+where
+    N: AstNode,
+{
+    fn fmt(&self, node: &N, f: &mut PyFormatter) -> FormatResult<()> {
+        write!(f, [source_position(node.start())])?;
+        self.fmt_leading_comments(node, f)?;
+        self.fmt_node(node, f)?;
+        self.fmt_dangling_comments(node, f)?;
+        self.fmt_trailing_comments(node, f)?;
+        write!(f, [source_position(node.start())])
+    }
+
+    /// Formats the node without comments. Ignores any suppression comments.
+    fn fmt_node(&self, node: &N, f: &mut PyFormatter) -> FormatResult<()> {
+        self.fmt_fields(node, f)?;
+
+        Ok(())
+    }
+
+    /// Formats the node's fields.
+    fn fmt_fields(&self, item: &N, f: &mut PyFormatter) -> FormatResult<()>;
+
+    /// Formats the [leading comments](rome_formatter::comments#leading-comments) of the node.
```

Comment:
Nit: This link won't work. You can point to the `crate::comments` instead
```suggestion
    /// Formats the [leading comments](crate::comments#leading-comments) of the node.
```